sound/soc/sof/intel/hda-ipc.c

```diff
@@ -234,19 +234,21 @@ irqreturn_t hda_dsp_ipc_irq_handler(int irq, void *context)
 {
 	struct snd_sof_dev *sdev = context;
 	int ret = IRQ_NONE;
+	u32 irq_status = 0;
 
 	spin_lock(&sdev->hw_lock);
 
 	/* store status */
-	sdev->irq_status = snd_sof_dsp_read(sdev, HDA_DSP_BAR,
-					    HDA_DSP_REG_ADSPIS);
+	irq_status = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_ADSPIS);
+	dev_vdbg(sdev->dev, "irq handler: irq_status:0x%x\n", irq_status);
 
 	/* invalid message ? */
-	if (sdev->irq_status == 0xffffffff)
+	if (irq_status == 0xffffffff)
 		goto out;
 
 	/* IPC message ? */
-	if (sdev->irq_status & HDA_DSP_ADSPIS_IPC) {
+	if (irq_status & HDA_DSP_ADSPIS_IPC) {
+		sdev->irq_status = irq_status;
```

Comment:
Allow me to challenge this patch. The only time where sdev->irq_status is used is in hda_dsp_ipc_irq_thread(), where we do this:
````c
        /* here we handle IPC interrupts only */
	if (!(sdev->irq_status & HDA_DSP_ADSPIS_IPC))
		return IRQ_NONE;
````
So what does this fix improve?

Reply:
@plbossart I agree. I think both the sdev->irq_status saving and the check again in hda_dsp_ipc_irq_thread() are unnecessary.

Reply:
@plbossart 
Let us think about a situation here:
Let IRQ A to be the IPC IRQ we need to handle and wake a thread.
Let IRQ B some IOC IRQ from HDA.

**IRQ A comes**, in the handle `sdev->irq_status` is set to be some useful value and we wake up the thread.

Before we go to the check bellow
```
/* here we handle IPC interrupts only */
	if (!(sdev->irq_status & HDA_DSP_ADSPIS_IPC))
		return IRQ_NONE;
```
**IRQ B comes** and reset the `sdev->irq_status` and do nothing

What will happen to the thread that already wake up by IRQ A?

It will just `return IRQ_NONE;`
And we will lost the chance to handle the IPC with IRQ A.

@ranj063 
You mean we did not check the status again in thread if we return the right value in handler?
That will only work if the handler will only wake the thread bind to itself. I am not sure if the code will do like that. If so I agree with you change to leave the check in one place.

https://github.com/thesofproject/linux/blob/90860434f04773357e5e55726bafd0a0a72430c5/sound/soc/sof/intel/hda.c#L509-L522